datastructures/clockchain.py

```diff
@@ -32,6 +37,81 @@ def __init__(self):
         genesis_dict = self.json_tick_to_chain_tick(tick)
         self.chain.put(genesis_dict)
 
+    def fork_pool_size(self):
+        return len(self.fork_pool)
+
+    def copy_chain(self, peer_addr):
+        logger.debug("Requesting chain from peer " + str(peer_addr))
+        peer_url = self.networker.reverse_peers.get(peer_addr, None)
+
+        if not peer_url:
+            logger.debug("Peer unknown, aborting chain request")
+            return False
+
+        logger.debug("Requesting chain from netloc " + str(peer_url))
+        r = requests.get(str(peer_url) + "/info/clockchain")
+
+        try:
+            alt_chain = r.json()['chain']
+        except KeyError:
+            logger.debug("Received invalid response from " + peer_addr + "; could not copy chain")
+            return False
+
+        alt_chain_is_valid = validate_clockchain(alt_chain)
+
+        if alt_chain_is_valid:
+            logger.debug("Received valid chain with " + str(len(alt_chain)) + " ticks")
+            logger.debug("Clearing current chain")
+            with self.chain.mutex:
+                self.chain.queue.clear()
+            logger.debug("Copying over new chain")
+            for tick in alt_chain:
```

Comment:
i think you're gonna want to include the copying inside the mutex as well, all the way down to self.chain.put(tick). Then you know for sure nobody else is messing with the chain while you update it

Reply:
I tried that, but that doesn't work; freezes the thread. Apparently we don't understand how this mutex works. Maybe it just needs a thread lock?